src/envs/pokemon_env/server/pokemon_environment.py

```diff
@@ -0,0 +1,691 @@
+            self._battle_task = None
+
+    def _extract_field_conditions(self, battle) -> Dict[str, Any]:
+        """Extract field conditions from battle state."""
+        conditions = {}
+
+        # Weather
+        if hasattr(battle, 'weather') and battle.weather:
+            for weather, turn_started in battle.weather.items():
+                conditions["weather"] = str(weather.name)
+                conditions["weather_turn"] = turn_started
+                break  # Only one weather active
+
+        # Terrain/Fields
+        if hasattr(battle, 'fields') and battle.fields:
+            terrains = []
+            for field, turn_started in battle.fields.items():
+                terrains.append({
+                    "name": str(field.name),
+                    "turn_started": turn_started
+                })
+            conditions["terrains"] = terrains
+
+        # Side conditions (your side)
+        if hasattr(battle, 'side_conditions'):
+            side_conds = {}
+            for condition, value in battle.side_conditions.items():
+                side_conds[str(condition.name)] = value
+            conditions["side_conditions"] = side_conds
+
+        # Opponent side conditions
+        if hasattr(battle, 'opponent_side_conditions'):
+            opp_side_conds = {}
+            for condition, value in battle.opponent_side_conditions.items():
+                opp_side_conds[str(condition.name)] = value
+            conditions["opponent_side_conditions"] = opp_side_conds
+
+        return conditions
+
+    def _compute_reward(self, battle, done: bool) -> float:
+        """Compute reward based on reward_mode."""
+
+        if self.reward_mode == "sparse":
+            # Only reward at end
+            if not done:
+                return 0.0
+
+            if battle.won:
+                return 1.0
+            elif battle.lost:
+                return -1.0
+            else:
+                return 0.0  # Tie
+
+        elif self.reward_mode == "dense":
+            # Per-turn reward shaping
+            reward = 0.0
+
+            # Reward for fainting opponent Pokemon
+            opponent_fainted = sum(1 for p in battle.opponent_team.values() if p.fainted)
+            new_faint_count = opponent_fainted - self._last_opponent_fainted
+            reward += new_faint_count * 0.2
+            self._last_opponent_fainted = opponent_fainted
+
+            # Penalty for losing own Pokemon
+            player_fainted = sum(1 for p in battle.team.values() if p.fainted)
+            new_player_faint = player_fainted - self._last_player_fainted
+            reward -= new_player_faint * 0.2
+            self._last_player_fainted = player_fainted
+
+            # Small reward for opponent HP damage
+            if battle.opponent_active_pokemon and hasattr(battle.opponent_active_pokemon, 'current_hp_fraction'):
+                current_hp = battle.opponent_active_pokemon.current_hp_fraction
+                if current_hp is not None:
```

Comment:
Also HP are often recovered. Moves like Recover, etc. There should be another reward for recovery of HP